File: src/pomodoro/widget/TextPanelWithIcon.java
package pomodoro.widget;

import com.intellij.openapi.wm.impl.status.TextPanel;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import javax.swing.*;
import java.awt.*;

/**
 * Based on {@link com.intellij.openapi.wm.impl.status.TextPanel.WithIconAndArrows}
 */
class TextPanelWithIcon extends TextPanel {
	private final static int GAP = 2;
	@Nullable
	private Icon myIcon;

	@Override
	protected void paintComponent(@NotNull final Graphics g) {
		super.paintComponent(g);
		if (getText() == null) return;

		Rectangle r = getBounds();
		Insets insets = getInsets();
		if (myIcon != null) {
			myIcon.paintIcon(this, g,
					insets.left - GAP - myIcon.getIconWidth(),
					r.height / 2 - myIcon.getIconHeight() / 2
			);
		}
	}

	@NotNull
	@Override
	public Insets getInsets() {
		Insets insets = super.getInsets();
		if (myIcon != null) {
			insets.left += myIcon.getIconWidth() + GAP * 2;
		}
		return insets;
	}

	@Override
	public Dimension getPreferredSize() {
		Dimension preferredSize = super.getPreferredSize();
		int deltaWidth = 0;
		if (myIcon != null) {
			deltaWidth += myIcon.getIconWidth();
		}
		return new Dimension(preferredSize.width + deltaWidth, preferredSize.height);
	}

	public void setIcon(@Nullable Icon icon) {
		myIcon = icon;
	}
}

File: src/pomodoro/widget/TextPanelWithIcon.kt
package pomodoro.widget

import com.intellij.openapi.wm.impl.status.TextPanel
import java.awt.Graphics
import java.awt.Insets
import javax.swing.Icon

/**
 * Based on [com.intellij.openapi.wm.impl.status.TextPanel.WithIconAndArrows]
 */
internal class TextPanelWithIcon: TextPanel() {
    private val gap = 2

    var icon: Icon? = null

    override fun paintComponent(g: Graphics) {
        super.paintComponent(g)
        if (text == null) return

        icon?.paintIcon(
            this, g,
            insets.left - gap - iconWidth(),
            bounds.height / 2 - iconWidth() / 2
        )
    }

    override fun getInsets(): Insets {
        val insets = super.getInsets()
        insets.left += iconWidth() + gap * 2
        return insets
    }

    private fun iconWidth() = icon?.iconWidth ?: 0
}
